cpp/src/fil/treelite_import.cu

```diff
@@ -69,32 +69,64 @@ int tree_root(const tl::Tree<T, L>& tree)
   return 0;  // Treelite format assumes that the root is 0
 }
 
-template <typename T, typename L>
-inline int max_depth(const tl::Tree<T, L>& tree)
+using empty = struct {
+};
+
+template <typename T,
+          typename L,
+          typename DescentAccumulator,
+          typename InnerFunc,
+          typename LeafFunc>
+inline void walk_tree(const tl::Tree<T, L>& tree,
+                      DescentAccumulator desc_acc,
+                      InnerFunc inner_func,
+                      LeafFunc leaf_func)
 {
   // trees of this depth aren't used, so it most likely means bad input data,
   // e.g. cycles in the forest
-  const int DEPTH_LIMIT = 500;
-  int root_index        = tree_root(tree);
-  typedef std::pair<int, int> pair_t;
-  std::stack<pair_t> stack;
-  stack.push(pair_t(root_index, 0));
-  int max_depth = 0;
+  using stackable = struct {
+    int id;
+    DescentAccumulator desc_acc;
+  };
+  std::stack<stackable> stack;
+  stack.push(stackable{tree_root(tree), desc_acc});
   while (!stack.empty()) {
-    const pair_t& pair = stack.top();
-    int node_id        = pair.first;
-    int depth          = pair.second;
+    stackable node = stack.top();
     stack.pop();
-    while (!tree.IsLeaf(node_id)) {
-      stack.push(pair_t(tree.LeftChild(node_id), depth + 1));
-      node_id = tree.RightChild(node_id);
-      depth++;
-      ASSERT(depth < DEPTH_LIMIT, "depth limit reached, might be a cycle in the tree");
+    while (!tree.IsLeaf(node.id)) {
+      DescentAccumulator left_acc, right_acc;
+      if constexpr (std::is_empty<DescentAccumulator>()) {
+        inner_func(node.id);
+      } else {
+        std::tie(left_acc, right_acc) = inner_func(node.id, node.desc_acc);
```

Comment:
I think that it can be more readable of `inner_func` accepts `left_acc` and `right_acc` as parameters passed by reference (or by pointer), but I'm not sure here.